File: lambda/generate_weekly_report/handler.py
"""Lambda function to generate weekly reports."""

import json
import logging
import os
from datetime import datetime, timedelta

from bedrock_client import BedrockClient
from dynamodb_client import DynamoDBClient
from s3_client import S3Client
from markdown_utils import create_weekly_report_document, get_week_string

logger = logging.getLogger()
logger.setLevel(logging.INFO)


def lambda_handler(event, context):
    """Generate a weekly report of PKM activity.

    Args:
        event: Step Function input or EventBridge scheduled event
        context: Lambda context

    Returns:
        Response dictionary
    """
    try:
        logger.info(f"Generating weekly report")

        # Allow date override from event (for testing)
        if event.get('date'):
            target_date = datetime.fromisoformat(event['date'])
        else:
            # Default to last week (report runs on Sunday for previous week)
            target_date = datetime.utcnow() - timedelta(days=7)

        week_str = get_week_string(target_date)
        logger.info(f"Target week: {week_str}")

        # Calculate week boundaries (Monday to Sunday)
        year, week, _ = target_date.isocalendar()
        week_start = datetime.strptime(f'{year}-W{week:02d}-1', '%Y-W%W-%w')
        week_end = week_start + timedelta(days=7)

        logger.info(f"Week range: {week_start} to {week_end}")

        # Initialize clients
        s3_client = S3Client()
        dynamodb_client = DynamoDBClient()
        bedrock_client = BedrockClient()

        # Query documents modified during the week
        logger.info(f"Querying documents modified since {week_start}")
        week_docs = dynamodb_client.get_documents_modified_since(week_start)

        # Filter documents within the target week
        week_docs_filtered = [
            doc for doc in week_docs
            if week_start.isoformat() <= doc.get('modified', '') < week_end.isoformat()
        ]

        logger.info(f"Found {len(week_docs_filtered)} documents for {week_str}")

        # Retrieve daily summaries for the week
        daily_summaries = []
        for i in range(7):
            day = week_start + timedelta(days=i)
            day_str = day.strftime('%Y-%m-%d')
            summary = s3_client.get_daily_summary(day_str)
            if summary:
                daily_summaries.append({
                    'date': day_str,
                    'content': summary
                })

        logger.info(f"Found {len(daily_summaries)} daily summaries")

        # Compile week data for analysis
        week_data = {
            'week': week_str,
            'start_date': week_start.strftime('%Y-%m-%d'),
            'end_date': (week_end - timedelta(days=1)).strftime('%Y-%m-%d'),
            'document_count': len(week_docs_filtered),
            'daily_summaries': daily_summaries,
            'documents': []
        }

        # Include sample of documents (up to 30)
        for doc in week_docs_filtered[:30]:
            doc_path = doc['PK'].replace('doc#', '')

            # Skip agent-generated documents
            if doc_path.startswith('_agent/'):
                continue

            week_data['documents'].append({
                'path': doc_path,
                'title': doc.get('title', 'Untitled'),
                'classification': doc.get('classification', 'unknown'),
                'tags': doc.get('tags', [])
            })

        # Classify documents by type for metrics
        classification_counts = {}
        for doc in week_docs_filtered:
            classification = doc.get('classification', 'unknown')
            classification_counts[classification] = classification_counts.get(classification, 0) + 1

        week_data['classification_counts'] = classification_counts

        logger.info(f"Generating report with {len(week_data['documents'])} documents")

        # Generate report using Bedrock
        model_id = os.environ.get('BEDROCK_MODEL_ID')
        report_content = bedrock_client.generate_weekly_report(week_data, model_id)

        # Create report document
        report_doc = create_weekly_report_document(
            week_str,
            report_content,
            len(week_docs_filtered)
        )

        # Upload report to S3
        report_key = s3_client.put_agent_output(
            'reports/weekly',
            f'{week_str}.md',
            report_doc
        )

        logger.info(f"Created weekly report: {report_key}")

        return {
            'statusCode': 200,
            'body': json.dumps({
                'week': week_str,
                'report_key': report_key,
                'document_count': len(week_docs_filtered),
                'daily_summaries_count': len(daily_summaries)
            })
        }

    except Exception as e:
        logger.error(f"Error generating weekly report: {e}", exc_info=True)
        return {
            'statusCode': 500,
            'body': json.dumps(f'Error: {str(e)}')
        }

File: lambda-bb/functions/generate_weekly_report/handler.clj
(ns handler
  "Lambda function to generate weekly reports of PKM activity"
  (:require [aws.s3 :as s3]
            [aws.dynamodb :as ddb]
            [aws.bedrock :as bedrock]
            [markdown.utils :as md]
            [cheshire.core :as json])
  (:import [java.time Instant ZonedDateTime ZoneOffset Duration DayOfWeek]
           [java.time.format DateTimeFormatter]
           [java.time.temporal ChronoUnit TemporalAdjusters]))

(def s3-bucket (System/getenv "S3_BUCKET_NAME"))
(def ddb-table (System/getenv "DYNAMODB_TABLE_NAME"))
(def bedrock-model (System/getenv "BEDROCK_MODEL_ID"))

(def ^:private date-formatter (DateTimeFormatter/ofPattern "yyyy-MM-dd"))

(defn- format-date
  "Format an Instant as yyyy-MM-dd"
  [inst]
  (.format (.atZone inst ZoneOffset/UTC) date-formatter))

(defn- parse-instant
  "Parse a date string to Instant"
  [date-str]
  (Instant/parse date-str))

(defn- minus-days
  "Subtract days from an Instant"
  [inst days]
  (.minus inst (Duration/ofDays days)))

(defn- plus-days
  "Add days to an Instant"
  [inst days]
  (.plus inst (Duration/ofDays days)))

(defn get-target-date
  "Get target date from event or default to last week"
  [event]
  (if-let [date-str (:date event)]
    (parse-instant date-str)
    ;; Default to last week (report runs on Sunday for previous week)
    (minus-days (Instant/now) 7)))

(defn calculate-week-boundaries
  "Calculate week start (Monday) and end (Sunday) for given date"
  [target-date]
  (let [zdt (.atZone target-date ZoneOffset/UTC)
        ;; Adjust to previous or same Monday
        week-start-zdt (.with zdt (TemporalAdjusters/previousOrSame DayOfWeek/MONDAY))
        week-start (.toInstant week-start-zdt)
        week-end (plus-days week-start 7)]
    {:start week-start
     :end week-end}))

(defn filter-week-docs
  "Filter documents to those within target week"
  [docs week-start-iso week-end-iso]
  (filter (fn [doc]
            (let [modified (or (:modified doc) "")]
              (and (>= (compare modified week-start-iso) 0)
                   (< (compare modified week-end-iso) 0))))
          docs))

(defn retrieve-daily-summaries
  "Retrieve daily summaries for the week"
  [week-start]
  (println "Retrieving daily summaries for the week")
  (let [summaries (atom [])]
    (doseq [i (range 7)]
      (let [day (plus-days week-start i)
            day-str (format-date day)
            summary (s3/get-daily-summary s3-bucket day-str)]
        (when summary
          (swap! summaries conj {:date day-str
                                :content summary}))))
    (println "Found" (count @summaries) "daily summaries")
    @summaries))

(defn compile-week-data
  "Compile week data for AI analysis"
  [week-str week-start week-end week-docs daily-summaries]
  (let [;; Calculate classification counts
        classification-counts (frequencies (map #(get % :classification "unknown") week-docs))

        ;; Sample documents (up to 30)
        sample-docs (take 30
                         (remove #(.startsWith (:PK %) "_agent/") week-docs))

        documents (mapv (fn [doc]
                         {:path (:PK doc)
                          :title (or (:title doc) "Untitled")
                          :classification (or (:classification doc) "unknown")
                          :tags (or (:tags doc) [])})
                       sample-docs)]

    {:week week-str
     :start_date (format-date week-start)
     :end_date (format-date (minus-days week-end 1))
     :document_count (count week-docs)
     :daily_summaries daily-summaries
     :documents documents
     :classification_counts classification-counts}))

(defn handler
  "Lambda handler for bblf runtime - receives raw HTTP request from Lambda Runtime API"
  [request]
  (try
    (let [event (json/parse-string (:body request) true)]
      (println "Generating weekly report")

      ;; Get target date
      (let [target-date (get-target-date event)
          week-str (md/get-week-string target-date)
          _ (println "Target week:" week-str)

          ;; Calculate week boundaries
          {:keys [start end]} (calculate-week-boundaries target-date)
          week-start start
          week-end end
          _ (println "Week range:" (str week-start) "to" (str week-end))

          ;; Query documents modified during the week
          week-start-iso (str week-start)
          week-end-iso (str week-end)
          _ (println "Querying documents modified since" week-start-iso)

          week-docs-all (ddb/get-documents-modified-since ddb-table week-start-iso :limit 2000)

          ;; Filter to target week
          week-docs (filter-week-docs week-docs-all week-start-iso week-end-iso)
          _ (println "Found" (count week-docs) "documents for" week-str)

          ;; Retrieve daily summaries
          daily-summaries (retrieve-daily-summaries week-start)

          ;; Compile week data
          week-data (compile-week-data week-str week-start week-end week-docs daily-summaries)

          _ (println "Generating report with" (count (:documents week-data)) "documents")

          ;; Generate report using Bedrock
          report-content (bedrock/generate-weekly-report bedrock-model week-data)

          ;; Create report document
          report-doc (md/create-weekly-report-document week-str
                                                       report-content
                                                       (count week-docs))

          ;; Upload report to S3
          report-key (s3/put-agent-output s3-bucket
                                         "reports/weekly"
                                         (str week-str ".md")
                                         report-doc)]

      (println "Created weekly report:" report-key)

      {:statusCode 200
       :body (json/generate-string {:week week-str
                              :report-key report-key
                              :document-count (count week-docs)
                              :daily-summaries-count (count daily-summaries)})}))

    (catch Exception e
      (println "Error generating weekly report:" (.getMessage e))
      (.printStackTrace e)
      {:statusCode 500
       :body (json/generate-string {:error (.getMessage e)})})))

;; For local testing
(defn -main [& args]
  (println "Running test weekly report generation")
  (println "Result:" (handler {} nil)))
